# Pandas



Pandas is a Python library used for working with datasets. It does that by helping us make sense of DataFrames, which are a form of two-dimensional structured data, like a table with columns and rows. But before we can do anything else, we need to start with data in a CSV file.

# Importing Data

## CSV Files



CSV stands for Comma Separated Values, and it's a file type that allows data to be saved in a table. Data presented in a table is called structured data, because it adheres to the idea that there is a meaningful relationship between the columns and rows. 

In [7]:
import pandas as pd 

df = pd.read_csv('data/iris.csv')

df.head()

,5.1,3.5,1.4,0.2,Iris-setosa
0,4.9,3.0,1.4,0.2,Iris-setosa
1,4.7,3.2,1.3,0.2,Iris-setosa
2,4.6,3.1,1.5,0.2,Iris-setosa
3,5.0,3.6,1.4,0.2,Iris-setosa
4,5.4,3.9,1.7,0.4,Iris-setosa


In [8]:
df.columns

Index(['5.1', '3.5', '1.4', '0.2', 'Iris-setosa'], dtype='object')

In [3]:
type(df)

pandas.core.frame.DataFrame

## Dictionaries

You can create a DataFrame from a Python dictionary using from_dict function.

In [4]:
import pandas as pd

data = {"col_1": [3, 2, 1, 0], "col_2": ["a", "b", "c", "d"]}
pd.DataFrame.from_dict(data)

,col_1,col_2
0,3,a
1,2,b
2,1,c
3,0,d


By default, DataFrame will be created using keys as columns. Note the length of the values should be equal for each key for the code to work. We can also let keys to be index instead of the columns:

In [5]:
pd.DataFrame.from_dict(data, orient="index")

,0,1,2,3
col_1,3,2,1,0
col_2,a,b,c,d


We can also specify column names:

In [4]:
pd.DataFrame.from_dict(data, orient="index", columns=["A", "B", "C", "D"])

,A,B,C,D
col_1,3,2,1,0
col_2,a,b,c,d


## JSON Files

JSON is short for JavaScript Object Notation. It is another widely used data format to store and transfer the data. It is light-weight and very human readable. In Python, we can use the json library to read JSON files. Here is an example of a JSON string.

In [10]:
info = """{
    "firstName": "Jane",
    "lastName": "Doe",
    "hobby": "running",
    "age": 35
}"""
print(info)

{
    "firstName": "Jane",
    "lastName": "Doe",
    "hobby": "running",
    "age": 35
}


Use json library to load the json string into a Python dictionary:

In [11]:
import json

data = json.loads(info)
data

{'firstName': 'Jane', 'lastName': 'Doe', 'hobby': 'running', 'age': 35}

We can load a json string or file into a dictionary because they are organized in the same way: key-value pairs.

In [7]:
data["firstName"]

'Jane'

A dictionary may not be as convenient as a DataFrame in terms of data manipulation and cleaning. But once we've turned our json string into a dictionary, we can transform it into a DataFrame using the from_dict method.

In [12]:
df = pd.DataFrame.from_dict(data, orient="index", columns=["subject 1"])
df

,subject 1
firstName,Jane
lastName,Doe
hobby,running
age,35


# Loading Compressed file in Python
In the big data era, it is very likely that we'll need to read data from compressed files. 

In [14]:
import pandas as pd 

df = pd.read_csv('data/public_company_bankruptcy_cases_2011.zip', compression='zip')

In [16]:
df

,DISTRICT,STATE,COMPANY NAME,ASSETS (MILLIONS),LIABILITIES (MILLIONS)
0,D,NV,"Ad Systems Communications, Inc.",405,4
1,D,NV,Alphatrade.com,686,4
2,SD,NY,AMBAC Financial Group,395,"1,683"
3,D,DE,"Ambassdors International , Inc.",86.44,87.32
4,D,NV,American Pacific Financial Corporation,19.18,161.08
...,...,...,...,...,...
62,NaN,NaN,NaN,NaN,NaN
63,NaN,NaN,NaN,NaN,NaN
64,NaN,NaN,NaN,NaN,NaN
65,NaN,NaN,NaN,NaN,NaN


In [18]:
df.columns

Index(['DISTRICT', 'STATE', 'COMPANY NAME', 'ASSETS (MILLIONS)',
       'LIABILITIES (MILLIONS)'],
      dtype='object')

In [12]:
df.head()

,DISTRICT,STATE,COMPANY NAME,ASSETS (MILLIONS),LIABILITIES (MILLIONS)
0,D,NV,"Ad Systems Communications, Inc.",405,4
1,D,NV,Alphatrade.com,686,4
2,SD,NY,AMBAC Financial Group,395,"1,683"
3,D,DE,"Ambassdors International , Inc.",86.44,87.32
4,D,NV,American Pacific Financial Corporation,19.18,161.08


## Pickle Files

The pickle module in Python is used for serializing and deserializing Python objects.

Serialization (also called pickling) is the process of converting a Python object in memory into a byte stream so that it can be stored on disk or transmitted over a network.

Deserialization (or unpickling) is the reverse operation — converting a byte stream back into a Python object in memory.

According to the official documentation, the following types of objects can be pickled:

- None
- Boolean values
- Numeric types: integers, floats, complex numbers
- Strings (both normal and Unicode)
- Collections such as tuples, lists, sets, and dictionaries (provided their contents are picklable)
- Functions defined at the top level of a module
- Built-in functions defined at the top level of a module
- Classes defined at the top level of a module



In [19]:
# Define a dictionary to store clothing items and their attributes (color, size)
clothes = {
    "shirt": ["red", "M"],
    "pant": ["yellow", "L"],
    "jacket": ["black", "L"]
}

# Display the dictionary
clothes


{'shirt': ['red', 'M'], 'pant': ['yellow', 'L'], 'jacket': ['black', 'L']}

In [20]:
import pickle

pickle.dump(clothes, open("./data/clothes.pkl", "wb"))

Now in the data folder, there will be a file named clothes.pkl. We can read the pickled file using the following code:

In [21]:
with open("./data/clothes.pkl", "rb") as f:
    unpickled = pickle.load(f)

In [22]:
unpickled

{'shirt': ['red', 'M'], 'pant': ['yellow', 'L'], 'jacket': ['black', 'L']}

Note first we are using wb inside the open function because we are creating this file, while deserializing the file, we are using rb to read the file

# Working with DataFrames



The first thing we need to do is import pandas; we'll use pd as an alias when we include it in our code.


Pandas is just a library; to get anything done, we need a dataset too. We'll use the read_csv method to create a DataFrame from a CSV file.

In [31]:
import pandas as pd


df = pd.read_csv("data/iris.csv",names=['Sepal_Length','Sepal_Width','Petal_Length','Petal_Width','Species'])
df.head()

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


# Working with DataFrame Indices

We can call the index column through .index:

In [24]:
df.index[:5]

RangeIndex(start=0, stop=5, step=1)

Use the set_index method, we can set the column department as the index instead. Note index column cannot have duplicate rows, like here we cannot set property_type as the index column.

In [25]:
df.set_index("Sepal_Length", inplace=True)
df.head()

,Sepal_Width,Petal_Length,Petal_Width,Species
Sepal_Length,,,,
5.1,3.5,1.4,0.2,Iris-setosa
4.9,3.0,1.4,0.2,Iris-setosa
4.7,3.2,1.3,0.2,Iris-setosa
4.6,3.1,1.5,0.2,Iris-setosa
5.0,3.6,1.4,0.2,Iris-setosa


Now you can see the index column has changed:

In [27]:
df['Sepal_Length']

KeyError: 'Sepal_Length'

Using the reset_index() function, we can reset index back to default integer counts, and department will become a column again.

In [29]:
df.reset_index(inplace=True)
df.head()

,index,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species
0,0,5.1,3.5,1.4,0.2,Iris-setosa
1,1,4.9,3.0,1.4,0.2,Iris-setosa
2,2,4.7,3.2,1.3,0.2,Iris-setosa
3,3,4.6,3.1,1.5,0.2,Iris-setosa
4,4,5.0,3.6,1.4,0.2,Iris-setosa


# Inspecting DataFrames

After creating a DataFrame, it is essential to inspect its contents to understand its structure and the type of data it holds.
Pandas provides several convenient methods for this purpose. In this section, we will focus on three commonly used ones: shape, info(), and head().

To examine the dimensionality of a DataFrame (i.e., the number of rows and columns), we can use the df.shape attribute.
Here’s an example:

In [22]:
df.shape

(150, 5)

The shape output tells us that the iris DataFrame has 150 rows and 5 columns.

If we're trying to get a general idea of what the DataFrame contained, we can use the info method. The code looks like this:

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Sepal_Length  150 non-null    float64
 1   Sepal_Width   150 non-null    float64
 2   Petal_Length  150 non-null    float64
 3   Petal_Width   150 non-null    float64
 4   Species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


The info() output provides a concise summary of the DataFrame. It displays the number of columns, their names, the data type of each column, and the count of non-null (non-missing) entries. This information is particularly useful for quickly assessing the structure and completeness of your dataset.

If we want to display all the rows in our DataFrame, we can use the print() function. However, this approach prints the entire dataset, which can be impractical for large DataFrames. While it may not pose an issue for small datasets, printing large ones can significantly slow down your workflow or overwhelm your console output.

A better approach is to preview only the first few rows using the head() method. By default, df.head() returns the first five rows of the DataFrame:

In [24]:
df.head()

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


By default, head returns the first five rows of data, but you can specify as many rows as you like. Here's what the code looks like for just the first two rows:

In [25]:
print(df.head(2))

   Sepal_Length  Sepal_Width  Petal_Length  Petal_Width      Species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa


# Sorting

We can sort the whole DataFrame by values of a column.

In [32]:
df.sort_values("Sepal_Width").head()

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species
60,5.0,2.0,3.5,1.0,Iris-versicolor
62,6.0,2.2,4.0,1.0,Iris-versicolor
119,6.0,2.2,5.0,1.5,Iris-virginica
68,6.2,2.2,4.5,1.5,Iris-versicolor
87,6.3,2.3,4.4,1.3,Iris-versicolor


We can also sort the DataFrame by it's index.

In [34]:
(df
 .set_index("Sepal_Length")
 .sort_index()
 .head()
 )

,Sepal_Width,Petal_Length,Petal_Width,Species
Sepal_Length,,,,
4.3,3.0,1.1,0.1,Iris-setosa
4.4,2.9,1.4,0.2,Iris-setosa
4.4,3.2,1.3,0.2,Iris-setosa
4.4,3.0,1.3,0.2,Iris-setosa
4.5,2.3,1.3,0.3,Iris-setosa


# Working with Columns


## Adding Columns

In [35]:
df["Sepal_Length2"] = df["Sepal_Length"]**2
df.head()

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species,Sepal_Length2
0,5.1,3.5,1.4,0.2,Iris-setosa,26.01
1,4.9,3.0,1.4,0.2,Iris-setosa,24.01
2,4.7,3.2,1.3,0.2,Iris-setosa,22.09
3,4.6,3.1,1.5,0.2,Iris-setosa,21.16
4,5.0,3.6,1.4,0.2,Iris-setosa,25.00


## Dropping Columns

In [36]:
df = df.drop("Sepal_Length2", axis="columns")
df.head()

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


Note that we specified that we wanted to drop a column by setting the axis argument to "columns". We can drop rows from the dataset if we change the axis argument to "index". If we wanted to drop row 2 from the df2 data, the code would look like this:

In [37]:
df2 = df.drop(1, axis="index")
df2.head()

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species
0,5.1,3.5,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa
5,5.4,3.9,1.7,0.4,Iris-setosa


## Adding Row

In [38]:
# Create a new row as a dictionary
new_row = {
    "Sepal_Length": 5.1,
    "Sepal_Width": 3.8,
    "Petal_Length": 1.5,
    "Petal_Width": 0.3,
    "Species": 'Iris-virginica'
}

# Add the new row using loc
df.loc[len(df)] = new_row

# Verify the addition
df.tail()

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species
146,6.3,2.5,5.0,1.9,Iris-virginica
147,6.5,3.0,5.2,2.0,Iris-virginica
148,6.2,3.4,5.4,2.3,Iris-virginica
149,5.9,3.0,5.1,1.8,Iris-virginica
150,5.1,3.8,1.5,0.3,Iris-virginica


## Dropping NaN



Including rows with empty cells can radically skew the results of our analysis, so we often drop them from the dataset.

In [39]:
import numpy as np
# Create a new row as a dictionary
new_row = {
    "Sepal_Length": np.nan,
    "Sepal_Width": np.nan,
    "Petal_Length": 3.2,
    "Petal_Width": np.nan,
    "Species": 'Iris-virginica'
}

# Add the new row using loc
df.loc[len(df)] = new_row

# Verify the addition
df.tail()

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species
147,6.5,3.0,5.2,2.0,Iris-virginica
148,6.2,3.4,5.4,2.3,Iris-virginica
149,5.9,3.0,5.1,1.8,Iris-virginica
150,5.1,3.8,1.5,0.3,Iris-virginica
151,NaN,NaN,3.2,NaN,Iris-virginica


In [40]:
print("df shape before dropping rows", df.shape)
df.dropna(inplace=True)
print("df shape after dropping rows", df.shape)
df.tail()

df shape before dropping rows (152, 5)
df shape after dropping rows (151, 5)


,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species
146,6.3,2.5,5.0,1.9,Iris-virginica
147,6.5,3.0,5.2,2.0,Iris-virginica
148,6.2,3.4,5.4,2.3,Iris-virginica
149,5.9,3.0,5.1,1.8,Iris-virginica
150,5.1,3.8,1.5,0.3,Iris-virginica


## Splitting Strings



It might be useful to split strings into their constituent parts, and create new columns to contain them. To do this, we’ll use the .str.split method, and include the character we want to use as the place where the data splits apart. In the colombia-real-estate-3 dataset, we might be interested breaking the "lat-lon" column into a "lat" column and a "lon" column. We’ll split it at “,” with code that looks like this:

In [41]:
df[["Flower", "Type"]] = df["Species"].str.split("-", expand=True)
df

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species,Flower,Type
0,5.1,3.5,1.4,0.2,Iris-setosa,Iris,setosa
1,4.9,3.0,1.4,0.2,Iris-setosa,Iris,setosa
2,4.7,3.2,1.3,0.2,Iris-setosa,Iris,setosa
3,4.6,3.1,1.5,0.2,Iris-setosa,Iris,setosa
4,5.0,3.6,1.4,0.2,Iris-setosa,Iris,setosa
...,...,...,...,...,...,...,...
146,6.3,2.5,5.0,1.9,Iris-virginica,Iris,virginica
147,6.5,3.0,5.2,2.0,Iris-virginica,Iris,virginica
148,6.2,3.4,5.4,2.3,Iris-virginica,Iris,virginica
149,5.9,3.0,5.1,1.8,Iris-virginica,Iris,virginica


## Recasting Data

In [42]:
print(df.info())
newdf = df.astype("str")
print(newdf.info())

<class 'pandas.core.frame.DataFrame'>
Index: 151 entries, 0 to 150
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Sepal_Length  151 non-null    float64
 1   Sepal_Width   151 non-null    float64
 2   Petal_Length  151 non-null    float64
 3   Petal_Width   151 non-null    float64
 4   Species       151 non-null    object 
 5   Flower        151 non-null    object 
 6   Type          151 non-null    object 
dtypes: float64(4), object(3)
memory usage: 9.4+ KB
None
<class 'pandas.core.frame.DataFrame'>
Index: 151 entries, 0 to 150
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Sepal_Length  151 non-null    object
 1   Sepal_Width   151 non-null    object
 2   Petal_Length  151 non-null    object
 3   Petal_Width   151 non-null    object
 4   Species       151 non-null    object
 5   Flower        151 non-null    object
 6   Type          151

In [43]:
df['Petal_Length2'] = (df.Petal_Length**2).astype(int)
df.head()

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species,Flower,Type,Petal_Length2
0,5.1,3.5,1.4,0.2,Iris-setosa,Iris,setosa,1
1,4.9,3.0,1.4,0.2,Iris-setosa,Iris,setosa,1
2,4.7,3.2,1.3,0.2,Iris-setosa,Iris,setosa,1
3,4.6,3.1,1.5,0.2,Iris-setosa,Iris,setosa,2
4,5.0,3.6,1.4,0.2,Iris-setosa,Iris,setosa,1


## Access a substring in a Series

In [44]:
df["Species"].str[0:5:2]

0      Ii-
1      Ii-
2      Ii-
3      Ii-
4      Ii-
      ... 
146    Ii-
147    Ii-
148    Ii-
149    Ii-
150    Ii-
Name: Species, Length: 151, dtype: object

## Replacing String Characters

In [45]:
df["Flower"] = df["Flower"].str.replace("ris", "RIS")
df.head()

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species,Flower,Type,Petal_Length2
0,5.1,3.5,1.4,0.2,Iris-setosa,IRIS,setosa,1
1,4.9,3.0,1.4,0.2,Iris-setosa,IRIS,setosa,1
2,4.7,3.2,1.3,0.2,Iris-setosa,IRIS,setosa,1
3,4.6,3.1,1.5,0.2,Iris-setosa,IRIS,setosa,2
4,5.0,3.6,1.4,0.2,Iris-setosa,IRIS,setosa,1


### Rename a column

In [47]:
df.rename(columns={"Type": "Kind"})

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species,Flower,Kind,Petal_Length2
0,5.1,3.5,1.4,0.2,Iris-setosa,IRIS,setosa,1
1,4.9,3.0,1.4,0.2,Iris-setosa,IRIS,setosa,1
2,4.7,3.2,1.3,0.2,Iris-setosa,IRIS,setosa,1
3,4.6,3.1,1.5,0.2,Iris-setosa,IRIS,setosa,2
4,5.0,3.6,1.4,0.2,Iris-setosa,IRIS,setosa,1
...,...,...,...,...,...,...,...,...
146,6.3,2.5,5.0,1.9,Iris-virginica,IRIS,virginica,25
147,6.5,3.0,5.2,2.0,Iris-virginica,IRIS,virginica,27
148,6.2,3.4,5.4,2.3,Iris-virginica,IRIS,virginica,29
149,5.9,3.0,5.1,1.8,Iris-virginica,IRIS,virginica,26


### Determine the unique values in a column

In [48]:
df["Species"].unique()

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

## Replacing Column Values

If you want to replace a columns' values, simply use the .replace() function:

In [49]:
df["Petal_Length2"].replace(1, 10)

0      10
1      10
2      10
3       2
4      10
       ..
146    25
147    27
148    29
149    26
150     2
Name: Petal_Length2, Length: 151, dtype: int64

In [50]:
replace_value = {10: 1, 2: 20}

df["Petal_Length2"].replace(replace_value)

0       1
1       1
2       1
3      20
4       1
       ..
146    25
147    27
148    29
149    26
150    20
Name: Petal_Length2, Length: 151, dtype: int64

# Concatenating

## Concatenating DataFrames



If we want to combine two DataFrames, we need to import Pandas and read in our data.

In [51]:
df1 = pd.read_csv("data/iris.csv",names=['Sepal_Length','Sepal_Width','Petal_Length','Petal_Width','Species'])
df2 = pd.read_csv("data/iris.csv",names=['Sepal_Length','Sepal_Width','Petal_Length','Petal_Width','Species'])
print("df1 shape:", df1.shape)
print("df2 shape:", df2.shape)

df1 shape: (150, 5)
df2 shape: (150, 5)


In [52]:
concat_df = pd.concat([df1, df2])
print("concat_df shape:", concat_df.shape)
concat_df.head()

concat_df shape: (300, 5)


,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


## Concatenating Series



We can also concatenate a Series using a similar set of commands. First, let's take two Series from the df1 and df2 respectively.

In [53]:
sr1 = df1["Petal_Length"]
sr2 = df2["Petal_Length"]
print("len sr1:", len(sr1)),
print(sr1.head())
print()
print("len sr2:", len(sr2)),
print(sr2.head())

len sr1: 150
0    1.4
1    1.4
2    1.3
3    1.5
4    1.4
Name: Petal_Length, dtype: float64

len sr2: 150
0    1.4
1    1.4
2    1.3
3    1.5
4    1.4
Name: Petal_Length, dtype: float64


In [54]:
concat_sr = pd.concat([sr1, sr2])
print("len concat_sr:", len(concat_sr)),
print(concat_sr.head())

len concat_sr: 300
0    1.4
1    1.4
2    1.3
3    1.5
4    1.4
Name: Petal_Length, dtype: float64


In [47]:
df

,Sepal_Length,Sepal_Width,Petal_Length,Petal_Width,Species,Flower,Type,Petal_Length2
0,5.1,3.5,1.4,0.2,Iris-setosa,IRIS,setosa,1
1,4.9,3.0,1.4,0.2,Iris-setosa,IRIS,setosa,1
2,4.7,3.2,1.3,0.2,Iris-setosa,IRIS,setosa,1
3,4.6,3.1,1.5,0.2,Iris-setosa,IRIS,setosa,2
4,5.0,3.6,1.4,0.2,Iris-setosa,IRIS,setosa,1
...,...,...,...,...,...,...,...,...
146,6.3,2.5,5.0,1.9,Iris-virginica,IRIS,virginica,25
147,6.5,3.0,5.2,2.0,Iris-virginica,IRIS,virginica,27
148,6.2,3.4,5.4,2.3,Iris-virginica,IRIS,virginica,29
149,5.9,3.0,5.1,1.8,Iris-virginica,IRIS,virginica,26


# Saving a DataFrame as a CSV

In [55]:
df.to_csv("data/iris_mod.csv",index=False)